physicsnemo/sym/eq/pde.py

```diff
@@ -90,52 +91,63 @@ def make_nodes(
         """
         nodes = []
+        node_dict = {}
+
         if create_instances == 1:
             if bool(freeze_terms):
                 print(
                     "Freezing of terms is not supported when create_instance = 1. No terms will be frozen!"
                 )
                 freeze_terms = {}  # override with an empty dict
             for name, eq in self.equations.items():
-                nodes.append(Node.from_sympy(eq, str(name), freeze_terms, detach_names))
+                node = Node.from_sympy(eq, str(name), freeze_terms, detach_names)
+                nodes.append(node)
+                if return_as_dict:
+                    node_dict[str(name)] = node
         else:
             # look for empty lists in freeze_terms dict
             for k in list(freeze_terms):
                 if not freeze_terms[k]:
                     freeze_terms.pop(k)
             for i in range(create_instances):
                 for name, eq in self.equations.items():
-                    if str(name) + "_" + str(i) in freeze_terms.keys():
-                        nodes.append(
-                            Node.from_sympy(
-                                eq,
-                                str(name) + "_" + str(i),
-                                freeze_terms[str(name) + "_" + str(i)],
-                                detach_names,
-                            )
+                    node_name = str(name) + "_" + str(i)
+                    if node_name in freeze_terms.keys():
+                        node = Node.from_sympy(
+                            eq,
+                            node_name,
+                            freeze_terms[node_name],
+                            detach_names,
                         )
                     else:
                         # set the freeze terms to an empty list
                         print(
                             "No freeze terms found for instance: "
-                            + str(name)
-                            + "_"
-                            + str(i)
+                            + node_name
                             + ", setting to empty"
                         )
-                        nodes.append(
-                            Node.from_sympy(
-                                eq,
-                                str(name) + "_" + str(i),
-                                [],
-                                detach_names,
-                            )
+                        node = Node.from_sympy(
+                            eq,
+                            node_name,
+                            [],
+                            detach_names,
                         )
-        return nodes
+                    nodes.append(node)
+                    if return_as_dict:
+                        node_dict[node_name] = node
+
+        if return_as_dict:
+            return node_dict
+        else:
+            return nodes
```

Comment:
Instead of building up *both* a list `nodes` and a dict `node_dict`, why not do this:

First, simply build up `node_dict` (dropping all references to `nodes`)

Then, at the end:
```python
if return_as_dict:
  return node_dict
else:
  return list(node_dict.values())
```

Note: since Python 3.9 (circa), dictionaries are insertion-order-preserving, so no list ordering issues on that front.

This change would simplify this method a lot, and remove duplicate data structures.